### PR TITLE
Bitbucket PAT support

### PR DESCRIPTION
**Pull request by imanuelcostigan**
Key user visible changes:
- `install_bitbucket()` authenticates using `username:pat` stored in `BITBUCKET_AUTH` envvar rather than using a username and password combination. This change breaks backward compatibiltiy 
- Bitbucket refs are now resolved in `install_bitbucket()` in a manner more consistent with `install_github()`.
- `github_resolve_ref` method renamed to `resolve_ref` to allow extension to non-Github methods

Other changes of note:
- Bitbucket `remote_package_name()` and `remote_sha()` methods no longer call `github_remote` class methods, but are now specifically defined for Bitbucket.
- `parse_git_repo()` renamed to `parse_github_repo` and refactored, making uses new `parse_repo()` function. The latter is also used by `parse_bitbucket_repo()`.
- Added Bitbucket test cases and updated github test cases where necessary.

Address #1157

### CODE REVIEW

**Comment by imanuelcostigan** [comment]
@jimhester I should also apologies for continued use of merge vs rebase. I'm not proficient enough to use rebase safely so I've stuck with merge. It doesn't look nice in project history, but I'd rather contribute than not.

**Comment by hadley** [comment]
@imanuelcostigan I'm merging pull requests with squashing so it doesn't really matter.

File: R/install-bitbucket.r [hunk 1]
```diff
@@ -3,36 +3,42 @@
 #' This function is vectorised so you can install multiple packages in
 #' a single command.
 #'
+#' To install from a private repo, or more generally, access the Bitbucket API
+#' with your own credentials, you will need to get an access token. You can
+#' create an access token following the instructions found in the
+#' \href{https://confluence.atlassian.com/bitbucket/app-passwords-828781300.html}{Bitbucket
+#' App Passwords documentation}. This PAT requires read-only access to your
+#' repositories and pull requests. Then store your Bitbucket user name and PAT
+#' separated by a colon in the environment variable \code{BITBUCKET_AUTH} (e.g.
```
Comment (jimhester) [inline R/install-bitbucket.r]: I think we should be consistent with the Github convention and call the environment variable BITBUCKET_PAT and function `bitbucket_pat()`

File: R/install-bitbucket.r [hunk 2]
```diff
@@ -56,20 +62,20 @@ remote_download.bitbucket_remote <- function(x, quiet = FALSE) {
...
   }
-
   dest <- tempfile(fileext = paste0(".zip"))
   src <- paste("https://bitbucket.org/", x$username, "/", tolower(x$repo), "/get/",
     x$ref, ".zip", sep = "")
-
-  if (!is.null(x$password)) {
+  if (!is.null(x$auth_token)) {
+    userpwd <- strsplit(x$auth_token, ":")[[1]]
+    stopifnot(!is.na(userpwd[2])) # stop if no colon in auth_token
```
Comment (jimhester) [inline R/install-bitbucket.r]: I think this test should be `length(userpwd) == 2` and we should have a more useful error message

```r
parts <- strsplit(x$auth_token, ":")[[1]]
if (length(parts) != 2) {
  stop ("`auth_token` needs a username and password separated by a colon", call. = FALSE)
}
```

File: tests/testthat/test-bitbucket.R [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  expect_equal(parse_bitbucket_repo("imanuelcostigan/devtest/somerepo@v1.0"),
+    list(username = "imanuelcostigan", repo = "devtest", subdir = "somerepo",
+      ref = "v1.0"))
+  expect_equal(parse_bitbucket_repo("imanuelcostigan/devtest/somerepo#1"),
+    list(username = "imanuelcostigan", repo = "devtest", subdir = "somerepo",
+      ref = bitbucket_pull("1")))
+})
+
+test_that("Bitbucket remote build works", {
+  # bitbucket_remote requires the OAuth token to be available in $HOME. So
```
Comment (jimhester) [inline tests/testthat/test-bitbucket.R]: This is no longer true right? Can we run these tests on the CIs now?

**Comment by imanuelcostigan** [comment]
@jimhester updated PR with changes per your feedback. Thanks!

**Comment by imanuelcostigan** [comment]
@hadley @jimhester I'm skipping remote build tests. Works locally for me, but not sure why this isn't working with_mock. Ready for merge from my end.

**Comment by imanuelcostigan** [comment]
@jimhester @hadley seeing as 1.12.0 has been released, should milestone for this be updated? When are you likely to merge this back in?

**Comment by kleinschmidt** [comment]
What's the status of this? Lots of our lab stuff is in bitbucket private repos so authenticating to bitbucket in a secure way is really important for us.

**Comment by jimhester** [comment]
This should be merged for the next release, which should be fairly soon, thank you everyone for the patience.

@kleinschmidt If you need to use this sooner you can always install this PR directly

```r
devtools::install_github("hadley/devtools#1220")
```

**Comment by kleinschmidt** [comment]
Ah, right, of course. Thanks, appreciate the work!

**Comment by imanuelcostigan** [comment]
@jimhester will this make it into the next update?

**Comment by sz-cgt** [comment]
Any news of  this?

**Comment by Serenthia** [comment]
Any news on when this is expected to be merged, as it missed the last release?  I'm still having to use devtools#1220, as I'm not committing plaintext creds in our projects.

Many thanks.

**Comment by paul4forest** [comment]
@Serenthia alternatively, you can use _install_git()_:

`devtools::install_git('ssh://git@bitbucket.org/user/repository.git')`

Make sure SSH support is enabled in git2r as explained in this [StackOverflow answer](https://stackoverflow.com/a/33620182/2641825). Also check Bitbucket instructions on how to [setup SSH for git](https://confluence.atlassian.com/bitbucket/set-up-ssh-for-git-728138079.html).

**Comment by hadley** [comment]
@imanuelcostigan would you be interested in porting this to remotes? That's what devtools will use for installation in the future. We are working on breaking devtools up into smaller pieces to try and make development a bit more fluid.

**Comment by imanuelcostigan** [comment]
@hadley I have implemented bitbucket interfaces into the ghit package https://github.com/cloudyr/ghit

I'll look to do the same for remotes package

**Comment by hadley** [comment]
Thanks!

**Pull request closed without merging**